substrate/frame/asset-rewards/src/lib.rs

```diff
@@ -0,0 +1,718 @@
+		NonExistentStaker,
+		/// An operation was attempted with a non-existent asset.
+		NonExistentAsset,
+		/// There was an error converting a block number.
+		BlockNumberConversionError,
+		/// The expiry block must be in the future.
+		ExpiryBlockMustBeInTheFuture,
+	}
+
+	/// Pallet's callable functions.
+	///
+	/// Allows optionally specifying an admin account for the pool. By default, the origin is made
+	/// admin.
+	#[pallet::call(weight(<T as Config>::WeightInfo))]
+	impl<T: Config> Pallet<T> {
+		/// Create a new reward pool.
+		#[pallet::call_index(0)]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			staked_asset_id: Box<T::AssetId>,
+			reward_asset_id: Box<T::AssetId>,
+			reward_rate_per_block: T::Balance,
+			expiry_block: BlockNumberFor<T>,
+			admin: Option<T::AccountId>,
+		) -> DispatchResult {
+			// Check the origin.
+			let creator = T::PermissionedOrigin::ensure_origin(origin.clone())?;
+
+			// Ensure the assets exist.
+			ensure!(
+				T::Assets::asset_exists(*staked_asset_id.clone()),
+				Error::<T>::NonExistentAsset
+			);
+			ensure!(
+				T::Assets::asset_exists(*reward_asset_id.clone()),
+				Error::<T>::NonExistentAsset
+			);
+
+			// Check the expiry block.
+			ensure!(
+				expiry_block > frame_system::Pallet::<T>::block_number(),
+				Error::<T>::ExpiryBlockMustBeInTheFuture
+			);
+
+			// Get the admin, defaulting to the origin.
+			let admin = admin.unwrap_or(creator.clone());
+
+			// Create the pool.
+			let pool = PoolInfoFor::<T> {
+				staked_asset_id: *staked_asset_id.clone(),
+				reward_asset_id: *reward_asset_id.clone(),
+				reward_rate_per_block,
+				total_tokens_staked: 0u32.into(),
+				reward_per_token_stored: 0u32.into(),
+				last_update_block: 0u32.into(),
+				expiry_block,
+				admin: admin.clone(),
+			};
+
+			// Insert it into storage.
+			let pool_id = NextPoolId::<T>::get();
+			Pools::<T>::insert(pool_id, pool);
+			NextPoolId::<T>::put(pool_id.saturating_add(1));
+
+			// Emit created event.
+			Self::deposit_event(Event::PoolCreated {
+				creator,
+				pool_id,
+				staked_asset_id: *staked_asset_id,
+				reward_asset_id: *reward_asset_id,
+				reward_rate_per_block,
+				expiry_block,
+				admin,
+			});
+
+			Ok(())
+		}
+
+		/// Stake tokens in a pool.
+		#[pallet::call_index(1)]
+		pub fn stake(origin: OriginFor<T>, pool_id: PoolId, amount: T::Balance) -> DispatchResult {
+			let caller = ensure_signed(origin)?;
+
+			// Always start by updating staker and pool rewards.
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			let staker_info = PoolStakers::<T>::get(pool_id, &caller).unwrap_or_default();
+			let (mut pool_info, mut staker_info) =
+				Self::update_pool_and_staker_rewards(&pool_info, &staker_info)?;
+
+			// Try to freeze the staker assets.
+			// TODO: (blocked https://github.com/paritytech/polkadot-sdk/issues/3342)
+
+			// Update Pools.
+			pool_info.total_tokens_staked.saturating_accrue(amount);
```

Comment:
Given the comment above, this could overflow since the TotalIssuance of an asset is not a cap on its Total locked amount. Maybe it shoudl rather use `checked_add` and return an error - especially when we want to make this permissionless and let ppl create pools for memcoins.